packages/editor/src/components/post-taxonomies/hierarchical-term-selector.js

```diff
@@ -464,18 +363,27 @@ class HierarchicalTermSelector extends Component {
 export default compose( [
 	withSelect( ( select, { slug } ) => {
 		const { getCurrentPost } = select( 'core/editor' );
-		const { getTaxonomy } = select( 'core' );
+		const { getTaxonomy, getEntityRecords } = select( 'core' );
+		const { isResolving } = select( 'core/data' );
 		const taxonomy = getTaxonomy( slug );
+		const availableTerms = getEntityRecords( 'taxonomy', slug, DEFAULT_QUERY );
+		const availableTermsTree = buildTermsTree( availableTerms );
 		return {
 			hasCreateAction: taxonomy ? get( getCurrentPost(), [ '_links', 'wp:action-create-' + taxonomy.rest_base ], false ) : false,
 			hasAssignAction: taxonomy ? get( getCurrentPost(), [ '_links', 'wp:action-assign-' + taxonomy.rest_base ], false ) : false,
 			terms: taxonomy ? select( 'core/editor' ).getEditedPostAttribute( taxonomy.rest_base ) : [],
+			isRequestingTerms: isResolving( 'core', 'getEntityRecords', [ 'taxonomy', slug, DEFAULT_QUERY ] ),
 			taxonomy,
+			availableTerms,
+			availableTermsTree,
 		};
 	} ),
-	withDispatch( ( dispatch ) => ( {
-		onUpdateTerms( terms, restBase ) {
-			dispatch( 'core/editor' ).editPost( { [ restBase ]: terms } );
+	withDispatch( ( dispatch, { slug, taxonomy } ) => ( {
+		onUpdateTerms( terms ) {
+			dispatch( 'core/editor' ).editPost( { [ taxonomy.rest_base ]: terms } );
+		},
+		addTermToEditedPost( term ) {
+			return dispatch( 'core/editor' ).addTermToEditedPost( slug, term );
```

Comment:
It's not entirely obvious whether `slug` here is slug of the term, or of the taxonomy.